File: android/app/src/main/java/com/masreferenceapp/storage/StorageSQLite.java
package com.masreferenceapp.storage;

import static android.content.Context.MODE_PRIVATE;
import static android.database.sqlite.SQLiteDatabase.openOrCreateDatabase;

import android.content.ContentValues;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.masreferenceapp.Status;

import java.io.File;


public class StorageSQLite extends ReactContextBaseJavaModule {
    ReactApplicationContext context;

    public StorageSQLite(ReactApplicationContext context) {
        super(context);
        this.context = context;
    }

    private void initDb(SQLiteDatabase db){
        // Table name and column names
        String TABLE_NAME = "myTable";
        String COLUMN_ID = "id";
        String COLUMN_NAME = "password";
        String CREATE_TABLE =
                "CREATE TABLE " + TABLE_NAME + "(" +
                        COLUMN_ID + " INTEGER PRIMARY KEY AUTOINCREMENT," +
                        COLUMN_NAME + " TEXT" +
                        ")";

        try {
            db.execSQL(CREATE_TABLE);
        } catch (SQLException e) {
            e.printStackTrace();
        }

    }

    private void deleteDb(){
        File dbpath = context.getDatabasePath("insecureSQLite.db");
        context.deleteDatabase(dbpath.getAbsolutePath());
    }

    @NonNull
    @Override
    public String getName() {

        return "StorageSQLite";

    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String createSQLiteDB(){

        File dbpath = context.getDatabasePath("insecureSQLite.db");
        SQLiteDatabase plainTextDb = openOrCreateDatabase(dbpath,null);
        this.initDb(plainTextDb);
        this.deleteDb();

        return Status.status("OK", "Message");
    }


    @ReactMethod(isBlockingSynchronousMethod = true)
    public String insertData(){

        File dbpath = context.getDatabasePath("insecureSQLite.db");
        SQLiteDatabase plainTextDb = openOrCreateDatabase(dbpath,null);
        this.initDb(plainTextDb);

        ContentValues values = new ContentValues();
        values.put("password", "SuperSecret");
        plainTextDb.insert("myTable", null, values);
        plainTextDb.close();

        this.deleteDb();

        return Status.status("OK", "Message");
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String updateData(){

        File dbpath = context.getDatabasePath("insecureSQLite.db");
        SQLiteDatabase plainTextDb = openOrCreateDatabase(dbpath,null);
        this.initDb(plainTextDb);

        ContentValues values = new ContentValues();
        values.put("password", "SuperSecret");
        long id = plainTextDb.insert("myTable", null, values);

        values.put("password", "newSecretPW");
        plainTextDb.update("myTable", values, "id" + " = ?", new String[]{String.valueOf(id)});

        this.deleteDb();

        return Status.status("OK", "Message");
    }


    @ReactMethod(isBlockingSynchronousMethod = true)
    public String deleteData(){

        File dbpath = context.getDatabasePath("insecureSQLite.db");
        SQLiteDatabase plainTextDb = openOrCreateDatabase(dbpath,null);
        this.initDb(plainTextDb);

        ContentValues values = new ContentValues();
        values.put("password", "SuperSecret");
        long id = plainTextDb.insert("myTable", null, values);


        plainTextDb.delete("myTable", "id" + " = ?",  new String[]{String.valueOf(id)});

        plainTextDb.close();
        this.deleteDb();

        return Status.status("OK", "Message");
    }

}

File: android/app/src/main/java/com/masreferenceapp/storage/StorageSQLite.kt
package com.masreferenceapp.storage

import android.content.ContentValues
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.masreferenceapp.ReturnStatus
import com.masreferenceapp.SensitiveData


class StorageSQLite(var context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    private fun initDb(db: SQLiteDatabase) {
        // Table name and column names
        val TABLE_NAME = "myTable"
        val COLUMN_ID = "id"
        val COLUMN_NAME = "password"
        val CREATE_TABLE = "CREATE TABLE " + TABLE_NAME + "(" +
                COLUMN_ID + " INTEGER PRIMARY KEY AUTOINCREMENT," +
                COLUMN_NAME + " TEXT" +
                ")"
        try {
            db.execSQL(CREATE_TABLE)
        } catch (e: SQLException) {
            e.printStackTrace()
        }
    }

    override fun getName(): String {
        return "StorageSQLite"
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun createSQLiteDB(): String {
        val dbpath = context.getDatabasePath("MasSqlite_"+(0..1000000).random()+".db")
        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
        initDb(plainTextDb)
        return ReturnStatus("OK", "Database created at: $dbpath").toJsonString()
    }

//    @ReactMethod(isBlockingSynchronousMethod = true)
//    fun insertData(): String {
//        val dbpath = context.getDatabasePath("insecureSQLite.db")
//        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
//        initDb(plainTextDb)
//        val values = ContentValues()
//        values.put("password", "SuperSecret")
//        plainTextDb.insert("myTable", null, values)
//        plainTextDb.close()
//        return ReturnStatus("OK", "Data inserted.").toJsonString()
//    }
//
//    @ReactMethod(isBlockingSynchronousMethod = true)
//    fun updateData(): String {
//        val dbpath = context.getDatabasePath("insecureSQLite.db")
//        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
//        initDb(plainTextDb)
//        val values = ContentValues()
//        values.put("password", "SuperSecret")
//        val id = plainTextDb.insert("myTable", null, values)
//        values.put("password", "newSecretPW")
//        plainTextDb.update("myTable", values, "id" + " = ?", arrayOf(id.toString()))
//        return ReturnStatus("OK", "Data updated.").toJsonString()
//    }
//
//    @ReactMethod(isBlockingSynchronousMethod = true)
//    fun deleteData(): String {
//        val dbpath = context.getDatabasePath("insecureSQLite.db")
//        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
//        initDb(plainTextDb)
//        val values = ContentValues()
//        values.put("password", "SuperSecret")
//        val id = plainTextDb.insert("myTable", null, values)
//        plainTextDb.delete("myTable", "id" + " = ?", arrayOf(id.toString()))
//        plainTextDb.close()
//        return ReturnStatus("OK", "Data deleted.").toJsonString()
//    }











    @ReactMethod(isBlockingSynchronousMethod = true)
    fun execSensitiveSQL(): String {
        val dbpath = context.getDatabasePath("MasSqlite_"+(0..1000000).random()+".db")
        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
        initDb(plainTextDb)

        plainTextDb.execSQL("INSERT INTO myTable (password) VALUES (\"" + SensitiveData.data + "\")")

        plainTextDb.close()

        val r = ReturnStatus("OK", "Using execSQ()L to insert sensitive data into DB stored at: $dbpath")
        return r.toJsonString()
    }


    @ReactMethod(isBlockingSynchronousMethod = true)
    fun execSensitiveSQLStoredProcedures(): String {
        val dbpath = context.getDatabasePath("MasSqlite_"+(0..1000000).random()+".db")
        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
        initDb(plainTextDb)

        val insertQuery ="INSERT INTO myTable (password) VALUES (?)"

        plainTextDb.execSQL(insertQuery, arrayOf(SensitiveData.data))

        plainTextDb.close()

        val r = ReturnStatus("OK", "Using execSQ()L to insert sensitive data into DB stored at: $dbpath")
        return r.toJsonString()
    }
        
    @ReactMethod(isBlockingSynchronousMethod = true)
    fun insertSensitive(): String {
        val dbpath = context.getDatabasePath("MasSqlite_"+(0..1000000).random()+".db")
        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
        initDb(plainTextDb)

        val values = ContentValues()
        values.put("password", SensitiveData.data)

        plainTextDb.insert("myTable", null, values)

        plainTextDb.close()

        val r = ReturnStatus("OK", "Using insert() to insert sensitive data into DB stored at: $dbpath")
        return r.toJsonString()
    }


    @ReactMethod(isBlockingSynchronousMethod = true)
    fun replaceSensitive(): String {
        val dbpath = context.getDatabasePath("MasSqlite_"+(0..1000000).random()+".db")
        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
        initDb(plainTextDb)

        plainTextDb.execSQL("INSERT INTO myTable (password) VALUES (\"DummyData\")")
        val newValues = ContentValues()
        newValues.put("password", SensitiveData.data)

        plainTextDb.replace("myTable", null, newValues)

        plainTextDb.close()

        val r = ReturnStatus("OK", "Using replace() to insert sensitive data into DB stored at: $dbpath")
        return r.toJsonString()
    }

        
    @ReactMethod(isBlockingSynchronousMethod = true)
    fun updateSensitive(): String {
        val dbpath = context.getDatabasePath("MasSqlite_"+(0..1000000).random()+".db")
        val plainTextDb = SQLiteDatabase.openOrCreateDatabase(dbpath, null)
        initDb(plainTextDb)

        plainTextDb.execSQL("INSERT INTO myTable (password) VALUES (\"DummyData\")")

        val newValues = ContentValues()
        newValues.put("password", SensitiveData.data)

        val selection: String = "id = ?"
        val selectionArgs: Array<String> = arrayOf("1")

        plainTextDb.update("myTable", newValues, selection, selectionArgs)

        plainTextDb.close()

        val r = ReturnStatus("OK", "Using update() to insert sensitive data into DB stored at: $dbpath")
        return r.toJsonString()
    }

        


    //@method
}